LibreNMS/Modules/Slas.php

```diff
@@ -0,0 +1,614 @@
+<?php
+/**
+ * SLA.php
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * @link       https://www.librenms.org
+ */
+
+namespace LibreNMS\Modules;
+
+use App\Models\Sla;
+use LibreNMS\Interfaces\Module;
+use LibreNMS\OS;
+use LibreNMS\RRD\RrdDefinition;
+use LibreNMS\Util\IP;
+
+class Slas implements Module
+{
+    //use SyncsModels;
+
+    /**
+     * Discover this module. Heavier processes can be run here
+     * Run infrequently (default 4 times a day)
+     *
+     * @param OS $os
+     */
+    public function discover(OS $os)
+    {
+        $device = $os->getDeviceArray();
+
+        if ($device['os_group'] == 'cisco') {
+            $this->discoverSlas($device);
+        } elseif ($device['os'] == 'junos') {
+            $this->discoverRpms($device);
+        }
+        // } else if ($device['os'] == 'huawei')
+        // {
+        //    $this->discoverHuaweis($device);
+        // }
+    }
+
+    /**
+     * Poll data for this module and update the DB / RRD.
+     * Try to keep this efficient and only run if discovery has indicated there is a reason to run.
+     * Run frequently (default every 5 minutes)
+     *
+     * @param OS $os
+     */
+    public function poll(OS $os)
+    {
+        $device = $os->getDeviceArray();
+
+        if ($device['os_group'] == 'cisco') {
```

Comment:
The best way to implement this is inheritance.  Look at some of the other modules.
There is an interface on the OS class that is called to poll something.  If the OS needs to override the default discovery/polling, just override that method in that OS's class file.

When multiple os use the same oids, we can use a trait (such as HostResources.php) to share the code.

Reply:
@geg347 : you can check how Mempools module does with the inheritance. That way, dedicated OS code ends up in OS class.

Reply:
Alright I will check that tomorrow.
For Juniper RPM the mib used in discovery seems standard and could work for other vendors. But for polling there are 3 mibs with one specific for Juniper.
I will check Mempools, thanks!